### PR TITLE
MSGraph Identity and Access

### PR DESCRIPTION
**Pull request by jochman**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation 

## TODO:
- [ ] Test Playbook
- [ ] Documentation

### CODE REVIEW

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml [hunk 1]
```diff
@@ -0,0 +1,165 @@
+category: Utilities
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml:1]: update category

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess_description.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
+## Authorization
+In both options below, the [device authorization grant flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-device-code) is used.
+
+In order to connect to the Azure Network Security Group using either Cortex XSOAR Azure App or the Self-Deployed Azure App:
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess_description.md:4]: update integration name

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+        Results to post in demisto
+    """
+    limit_str = args.get('limit', '')
+    try:
+        limit = int(limit_str)
+    except ValueError:
+        raise DemistoException(f'Limit must be an integer, not "{limit_str}"')
+    results = ms_client.get_directory_roles(limit)
+    return CommandResults(
+        "Graph.Identity.Role",
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:182]:
```suggestion
        "MSGraphIdentity.Role",
```

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+    Returns:
+        Results to post in demisto
+    """
+    limit_str = args.get('limit', '')
+    try:
+        limit = int(limit_str)
+    except ValueError:
+        raise DemistoException(f'Limit must be an integer, not "{limit_str}"')
+    results = ms_client.get_directory_roles(limit)
+    return CommandResults(
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:181]: pass tabletomarkdown with headrs and removenull

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    role_id = args['role_id']
+    if results := ms_client.get_role_members(role_id, limit):
+        context = [member['id'] for member in results]
+        return CommandResults(
+            "Graph.Identity.RoleMember",
+            "id",
+            outputs=context,
+            raw_response=results,
+            readable_output=tableToMarkdown(
+                f"Role members of role id '{role_id}'", context, ["id"]
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:212]:
```suggestion
                f"Role '{role_id}' members", context, ["id"]
```

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        type: String
+      - contextPath: Graph.Identity.Role.roleTemplateId
+        description: The id of the directoryRoleTemplate that this role is based on.
+        type: String
+  - deprecated: false
+    description: Activate a role by its template id.
+    execution: false
+    name: msgraph-identity-directory-role-activate
+    arguments:
+      - name: role_template_id
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml:78]: for all id args mention how to retrieve them

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        secret: false
+  isfetch: false
+  runonce: false
+  script: '-'
+  type: python
+  subtype: python3
+  dockerimage: demisto/crypto:1.0.0.14297
+fromversion: 5.0.0
+tests:
+- No tests (auto formatted)
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml:165]: add test playbook

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    role_object_id = args['role_id']
+    client.add_member_to_role(role_object_id, user_id)
+    return f"User ID {user_id} has been added to role {role_object_id}"
+
+
+def remove_member_from_role(client: Client, args: dict) -> str:
+    role_object_id = args['role_id']
+    user_id = args['user_id']
+    client.remove_member_from_role(role_object_id, user_id)
+    return f"User ID {user_id} has been remove from role {role_object_id}"
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:244]:
```suggestion
    return f"User ID {user_id} has been removed from role {role_object_id}"
```

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+
+        Returns:
+            directoryRole object.
+
+        Docs:
+            https://docs.microsoft.com/en-us/graph/api/directoryrole-post-directoryroles?view=graph-rest-1.0&tabs=http
+        """
+        return self.ms_client.http_request(
+            'POST',
+            'v1.0/roleTemplateId',
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:78]:
```suggestion
            'v1.0/directoryRoles',
```

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+            https://docs.microsoft.com/en-us/graph/api/directoryrole-list?view=graph-rest-1.0&tabs=http
+        """
+        results = list()
+        res = self.ms_client.http_request(
+            'GET', 'v1.0/directoryRoles')
+        results.extend(res.get('value'))
+        while (next_link := res.get('@odata.nextLink')) and len(results) < limit:
+            res = self.ms_client.http_request('GET', '', next_link)
+            results.extend(res.get('value'))
+        return results[:limit]
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:53]: please use `top` here instead

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+            limit: Maximum roles to get.
+
+        Returns:
+            directoryObject
+
+        Docs:
+            https://docs.microsoft.com/en-us/graph/api/directoryrole-list-members?view=graph-rest-1.0&tabs=http
+        """
+        return self.ms_client.http_request(
+            'GET', f'v1.0/directoryRoles/{role_id}/members')['value'][:limit]
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:69]: please use `top` here instead

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+            https://docs.microsoft.com/en-us/graph/api/directoryrole-post-directoryroles?view=graph-rest-1.0&tabs=http
+        """
+        try:
+            return self.ms_client.http_request(
+                'POST',
+                'v1.0/directoryRoles',
+                json_data={'roleTemplateId': template_id}
+            )
+        except Exception as exc:
+            if 'A conflicting object with one or more of the specified property values is present in the directory.' in str(exc):
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:89]: why are we modifying the exception? it seems indicative enough

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+            '@odata.id': f'https://graph.microsoft.com/v1.0/directoryObjects/{user_id}'
+        }
+        self.ms_client.http_request(
+            'POST',
+            f'v1.0/directoryRoles/{role_object_id}/members/$ref',
+            headers={'Content-Type': 'application/json'},
+            json_data=body,
+            return_empty_response=True
+        )
+        return True
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:119]: why are we returning anything here?

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+        Docs:
+            https://docs.microsoft.com/en-us/graph/api/directoryrole-post-members?view=graph-rest-1.0&tabs=http
+        """
+        body = {
+            '@odata.id': f'https://graph.microsoft.com/v1.0/directoryObjects/{user_id}'
+        }
+        self.ms_client.http_request(
+            'POST',
+            f'v1.0/directoryRoles/{role_object_id}/members/$ref',
+            headers={'Content-Type': 'application/json'},
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:115]: why are we passing `headers` here?

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 15]
```diff
@@ -0,0 +1,310 @@
...
+
+        Docs:
+            https://docs.microsoft.com/en-us/graph/api/directoryrole-delete-member?view=graph-rest-1.0&tabs=http
+        """
+        self.ms_client.http_request(
+            'DELETE',
+            f'v1.0/directoryRoles/{role_object_id}/members/{user_id}/$ref',
+            return_empty_response=True
+        )
+        return True
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:142]: why are returning anything here?

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 16]
```diff
@@ -0,0 +1,310 @@
...
+
+def test_connection(client: Client) -> str:
+    client.ms_client.get_access_token()
+    return '✅ Success!'
+
+
+def reset_auth() -> CommandResults:
+    set_integration_context({})
+    return CommandResults(
+        readable_output='Authorization was reset successfully. Run **!microsoft-teams-auth-start** to '
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:171]:
```suggestion
        readable_output='Authorization was reset successfully. Run **!msgraph-identity-auth-start** to '
```

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 17]
```diff
@@ -0,0 +1,310 @@
...
+        command = demisto.command()
+        params = demisto.params()
+        args = demisto.args()
+        client = Client(
+            app_id=params['app_id'],
+            verify=not params.get('insecure', False),
+            proxy=params.get('proxy', False),
+        )
+        if command == 'test-module':
+            return_results('The test module is not functional, run the msgraph-identity--auth-start command instead.')
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:277]:
```suggestion
            return_results('The test module is not functional, run the msgraph-identity-auth-start command instead.')
```

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py [hunk 18]
```diff
@@ -0,0 +1,310 @@
...
+    return f"User ID {user_id} has been removed from role {role_object_id}"
+
+
+def main():
+    demisto.debug(f'Command being called is {demisto.command()}')
+    try:
+        command = demisto.command()
+        params = demisto.params()
+        args = demisto.args()
+        client = Client(
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:271]: i think you need to call `handle_proxy` here

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess_description.md [hunk 19]
```diff
@@ -0,0 +1,21 @@
+## Authorization
+In both options below, the [device authorization grant flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-device-code) is used.
+
+In order to connect to the AMicrosoft Graph Identity & Access using either Cortex XSOAR Graph App or the Self-Deployed Graph App:
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess_description.md:4]:
```suggestion
In order to connect to the Microsoft Graph Identity & Access using either Cortex XSOAR Graph App or the Self-Deployed Graph App:
```

File: Packs/MicrosoftGraphIdentityandAccess/pack_metadata.json [hunk 20]
```diff
@@ -0,0 +1,15 @@
+{
+    "name": "Microsoft Graph Services",
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/pack_metadata.json:2]: update name, description and category
add tags and usecases if relevant

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py (hunk 11 above)
Comment (jochman) [reply Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:69]: top isn't supported

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py (hunk 10 above)
Comment (jochman) [reply Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.py:53]: top isn't supported

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+- display: Trust any certificate (not secure)
+  name: insecure
+  required: false
+  type: 8
+- display: Use system proxy settings
+  name: proxy
+  required: false
+  type: 8
+description: Use the Microsoft Graph Identity & Access integration to manage roles
+  and membersn.
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml:21]:
```suggestion
  and members.
```

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+  commands:
+  - deprecated: false
+    description: Run this command to start the authorization process and follow the
+      instructions in the command results.
+    execution: false
+    name: msgraph-identity-auth-start
+  - deprecated: false
+    description: |-
+      Run this command to complete the authorization process.
+      Should be used after running the msgraph-apps-auth-start command.
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml:34]:
```suggestion
      Should be used after running the msgraph-identity-auth-start command.
```

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+    - name: limit
+      default: false
+      description: Maximum results to fetch
+      isArray: false
+      required: false
+      defaultValue: 10
+      secret: false
+    outputs:
+    - contextPath: MSGraphIdentity.Role.deletedDateTime
+      description: Presents if role is deleted.
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml:60]: please update description here

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+    arguments:
+    - name: role_template_id
+      default: false
+      description: Role template to activate. Can be retrieved from msgraph-identity-directory-roles-list.
+      isArray: false
+      required: true
+      secret: false
+    outputs:
+    - contextPath: MSGraphIdentity.Role.deletedDateTime
+      description: Presents if role is deleted.
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess.yml:87]: please update description here

File: Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess_description.md [hunk 25]
```diff
@@ -0,0 +1,21 @@
+## Authorization
+In both options below, the [device authorization grant flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-device-code) is used.
+
+In order to connect to the Microsoft Graph Identity & Access using either Cortex XSOAR Graph App or the Self-Deployed Graph App:
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/Integrations/MicrosoftGraphIdentityandAccess/MicrosoftGraphIdentityandAccess_description.md:4]:
```suggestion
In order to connect to Microsoft Graph Identity & Access using either Cortex XSOAR Graph App or the Self-Deployed Graph App:
```

**Comment by ShirleyDenkberg** [comment]
@Itay4 Doc review completed.  The integration README needs to be generated.

File: Packs/MicrosoftGraphIdentityandAccess/pack_metadata.json [hunk 26]
```diff
@@ -0,0 +1,17 @@
+{
+    "name": "Microsoft Graph Identity & Access",
+    "description": "Use this pack to manage roles and members.",
```
Comment (Itay4) [inline Packs/MicrosoftGraphIdentityandAccess/pack_metadata.json:3]:
```suggestion
    "description": "Use this pack to manage roles and members in Microsoft.",
```

**Review by Itay4 — APPROVED** [review]

**Pull request merged**